CNV-4617 nmstate remove (#28420)
openshift/openshift-docs: modules/virt-removing-interface-from-nodes.adoc

[id="virt-removing-interface-from-nodes_{context}"]
= Removing an interface from nodes

> Remove an interface from nodes by editing the `NodeNetworkConfigurationPolicy` object and set the `state` of the interface to `absent`.

Reviewer: IIUC from the example later in the text, it's supposed to be :
```
Remove an interface from nodes by editing the `NodeNetworkConfigurationPolicy` object by setting the `state` of the interface to `absent`
```

Also, I think we should rephrase this sentence, as both this paragraph and the following one start with the same "Remove an interface"/"Removing an interface".

Author: Thanks Ram, updated mostly as suggested

> Removing an interface from a node does not automatically restore the node network configuration to a previous state. After you remove an interface, any of the node NICs throughout the cluster that were previously attached or subordinate to the interface are placed in a `down` state and become unreachable. +

Reviewer 2: The need to bring eth0 up again applies only when it was attached to a bridge or a bonding. Maybe we could change:

"subordinate to the interface" -> "subordinate to a bridge or a bonding"

Author: Updated as suggested. Do you think we need to be more explicit about the difference between bridge & bond vs vlan interface or do you think this is sufficient

> Removing an interface from a node does not automatically restore the node network configuration to a previous state. After you remove an interface, any of the node NICs throughout the cluster that were previously attached or subordinate to a bridge or bonding interface are placed in a `down` state and become unreachable. +

Author: Thanks Yossi. I've expanded this para and the previous para. Hopefully this is clearer:
"Removing an interface from a node does not automatically restore the node network configuration to a previous state. If you want to restore the previous state, you will need to define that node network configuration in the policy.

If you remove a bridge or bonding interface, any of the node NICs throughout the cluster that were previously attached or subordinate to that bridge or bonding interface are placed in a down state and become unreachable.
To maintain connectivity with the NIC, specify a NIC configuration in the same policy with a status of up and either DHCP or a static IP address."

> To maintain connectivity with the NIC, specify an active NIC configuration in the same policy.

Reviewer 3: What does `specify an active NIC configuration` mean? Is this a way of saying to enable dhcp?

Author: Intended as with a status of 'up'. I have now updated this to be more clear:
"To maintain connectivity with the NIC, specify a NIC configuration in the same policy with a status of `up` and either DHCP or a static IP address."
and in the procedure:
"The following example removes a Linux bridge. The policy also configures the `eth1` NIC with DHCP to avoid losing connectivity:"

Reviewer 3: Oh, I see. Thanks, it's crystal clear now.
/lgtm

[id="virt-removing-interface-from-nodes_{context}"]
= Removing an interface from nodes

> You can remove an interface from nodes by editing the `NodeNetworkConfigurationPolicy` object and setting the `state` of the interface to `absent`.

Author: Yes. Yes it does :) (more, that is)

> Removing an interface from a node does not automatically restore the node network configuration to a previous state. If you want to restore the previous state, you will need to define that node network configuration in the policy.

Author: No order of ops that I'm aware of. Basically you replace one policy with another. There's no such thing as removing the config in order to roll it back - you must update it to a new state, even if that state is the same as an old state.

> If you remove a bridge or bonding interface, any of the node NICs throughout the cluster that were previously attached or subordinate to that bridge or bonding interface are placed in a `down` state and become unreachable. +

Reviewer 4: Suggest removing `+`. I think it would look better with the regular line break.

> If you remove a bridge or bonding interface, any of the node NICs throughout the cluster that were previously attached or subordinate to that bridge or bonding interface are placed in a `down` state and become unreachable. +

Author: Great suggestion

> To maintain connectivity with the NIC, specify a NIC configuration in the same policy with a status of `up` and either DHCP or a static IP address.

Reviewer 4: This sentence is a little unclear to me. Some potentially ill-advised suggestions:

s/with the NIC/with the NIC that you removed (? I'm not sure which NIC you're referring to here.)
s/specify a NIC configuration/configure a NIC
s/with a status/that has a status
s/and either/that uses either

Author: NIC here refers to the node NIC that was previously attached or subordinate to the removed bridged/bonded interface. With your feedback in mind I've removed the line break so that it is one paragraph and modified that last sentence:
To avoid losing connectivity, configure the node NIC in the same policy so that it has a status of `up` and either DHCP or a static IP address.

.Procedure

> . Update the `NodeNetworkConfigurationPolicy` manifest used to create the interface. The following example removes a Linux bridge. The policy also configures the `eth1` NIC with DHCP to avoid losing connectivity:

Reviewer 4: ```suggestion
. Update the `NodeNetworkConfigurationPolicy` manifest used to create the interface. The following example removes a Linux bridge and configures the `eth1` NIC with DHCP to avoid losing connectivity:
```

    node-role.kubernetes.io/worker: "" <3>
  desiredState:
    interfaces:
    - name: br1
      type: linux-bridge
      state: absent <4>
    - name: eth1
      type: ethernet <5>
      state: up <6>
      ipv4:
        dhcp: true <7>
        enabled: true <8>
----
<1> Name of the policy.
<2> Optional: If you do not include the `nodeSelector` parameter, the policy applies to all nodes in the cluster.
<3> This example uses the `node-role.kubernetes.io/worker: ""` node selector to select all worker nodes in the cluster.
<4> Changing the state to `absent` removes the interface.

> <5> The type of interface. This example creates an Ethernet networking interface.

Reviewer 5: I would add an extra explanation line (sorry, I don't know what the official term is) for the eth1 interface - probably for the "name" entry - which say something like:
The name of the interface that is going to be unattached from the bridge interface.

I think it is important in order to clarify that eth1 is not some arbitrary node interface, because in this bridge teardown policy there's no link between the bridge and its ports (as opposed to the setup policy above, where eth1 is explicitly declared as a subordinate port of the bridge).

Author: We call them 'callouts' :)
Great point. I've added callout with description:
"The name of the interface that is to be unattached from the bridge interface."

> <6> The requested state for the interface after creation.

Reviewer 5: Not "after creation". This interface already exists, and this entry defined the requested state after detaching it from the master interface (bridge, bond, ...)

Author: Updated to:
The requested state for the interface.

<7> Optional: If you do not use `dhcp`, you can either set a static IP or leave the interface without an IP address.
<8> Enables `ipv4` in this example.

. Update the policy on the node and remove the interface:
+
[source,terminal]
----
$ oc apply -f <br1-eth1-policy.yaml> <1>
----

> 1> File name of the policy manifest.

Reviewer 5: Looks like the preceding &lt sign was removed by mistake, so in the build preview it appears as "1>" rather than "1" framed inside a circle.
